### DISCUSSION TITLE
Compiling PyGIMLi from Source on Apple M-Based Macs

### DISCUSSION THREAD

**Discussion by robinthibaut** [gimli-org/pyGIMLi#603] [Show and tell]
Hello everyone,

I've encountered some difficulties with the [instructions](https://www.pygimli.org/compilation.html) given for compiling PyGIMLi from source on Macs. They were a great starting point, but I had to work through some roadblocks to make it work. I'd like to share adjusted instructions that should provide a good starting point for others in the community. While some steps may need tweaking for your individual systems, this guide should give you a fair idea:

1. Install Homebrew:
  ```bash 
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install.sh)"
  ```

2. Install essential libraries:
  ```bash
brew install boost 
brew install cmake 
brew install git 
brew install wget
brew install mercurial
brew install xorgproto
  ```

3. Install Miniforge: Download the latest version from Conda-forge's Github page.
  ```bash
curl -L -O "https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-$(uname)-$(uname -m).sh"
bash Miniforge3-$(uname)-$(uname -m).sh
  ```

4. Prepare a Python environment with an M-optimized version of Python through this Miniforge Conda, ensure to install necessary python packages:
  ```bash
conda create -n pygimli_env python=3.9
conda activate pygimli_env
conda install -c conda-forge boost numpy scipy matplotlib
  ```

5. Clone the PyGIMLi repository, create a build directory, and build PyGIMLi:
  ```bash
mkdir -p ~/src/gimli
cd ~/src/gimli
git clone https://github.com/gimli-org/gimli.git
cd gimli/
mkdir build
cd build/
  ```

Importantly, the paths listed in the CMake command need to correspond to the version of Python being used, which you can check using `which python3` and `python3 --version`
  ```bash
PYTHON_EXEC=$(which python3)
PYTHON_INC=$(python3 -c 'import sysconfig; print(sysconfig.get_path("include"))')
PYTHON_LIB=$(python3-config --configdir)

export CPLUS_INCLUDE_PATH=$PYTHON_INC

cmake .. -DPYTHON_EXECUTABLE=$PYTHON_EXEC -DPYTHON_LIBRARY=$PYTHON_LIB -DPYTHON_INCLUDE_DIR=$PYTHON_INC 
make -j 8
make pygimli
  ```

You can test your installation with:

  ```bash
python3 -c 'import pygimli as pg; print(pg.__version__)'
  ```

**Watch out: You might notice that PyGIMLi creates an .so file named `lib_pygimli_.so`, make sure to rename it to `_pygimli_.so` since `_pygimli_.so` is the name PyGIMLi internally uses.**

I hope this guide can be of some use.

Cheers,
Robin

**Comment by halbmy** [comment]
Dear Robin,
thank you so much for sharing your findings with the community! Maybe there are (except @florian-wagner who generates the conda packages) others around to test and share their experience. Would be interesting to found out whether this depends on the OS version, kernel, chip etc. 

Good to have the Mac users on board. Are the conda versions working for you?

**Reply by robinthibaut** [reply]
Dear Thomas,

I attempted to install pygimli using conda on my M2 Mac as per the instructions, but I encountered a couple of issues when running `conda create -n pg -c gimli -c conda-forge pygimli=1.4.3`.

First, there was a problem loading the `conda-libmamba-solver` plugin due to a missing library, specifically `libarchive.13.dylib`.

Second, I had trouble finding the `tetgen` package in the channels `gimli` and `conda-forge`. From the error messages, it seems `tetgen` might not be available for my specific architecture `osx-arm64`.

That's why I decided to compile it from source - it now works smoothly!
It would be nice for the community to have access to conda versions working out of the box though.  I'd be happy to help test different setups or solutions for Macs if anyone needs that.

Note that I also have access to an Intel-based Mac, on which the conda installation succeeded.

**Comment by carsten-forty2** [comment]
Hello Robin,

> Watch out: You might notice that PyGIMLi creates an .so file named lib_pygimli_.so, make sure to rename it to _pygimli_.so since _pygimli_.so is the name PyGIMLi internally uses.

ahh .. I found the same weird leading 'lib' also for the windows build, which probably comes from a recent cmake version. I recently added a special cmake target to rename it for the windows build and should add this target for the mac build to.

Thanks a lot for the building instructions .. we will add a link to this thread into the documentation, or can we copy it directly? 

Cheers,
Carsten

**Reply by robinthibaut** [reply]
Hello Carsten,
Yes, I felt really bummed out after getting the error ```no module named pgcore```, although compilation apparently ran successfully. Luckily, after some digging, I found out about the `.so` file.  Please use/copy/edit my instructions as you wish!
Cheers,
Robin

**Reply by carsten-forty2** [reply]
Bad me to try blaming cmake for my own faults. I messed up the cmake script and accidently removed the lib prefix. Its fixed now and the library output name is correct again for all platforms.

**Comment by ruboerner** [comment]
Hello all,

many thanks to @robinthibaut for sharing the installation steps for Mac OS!

I could successfully compile pygimli after some modifications to the guide [given above](https://github.com/gimli-org/gimli/discussions/603#discussion-5751447) :

1. As `umfpack` was missing in my `homebrew` installation, I had to add `suite-sparse` to the list of essential libraries:
```
homebrew install suite-sparse
```
2. Further, `cmake` struggled with finding `lapack` and `openblas`, so I had to modify the build commands (as per item 5 in the list above):
```
PYTHON_EXEC=$(which python3)
PYTHON_INC=$(python3 -c 'import sysconfig; print(sysconfig.get_path("include"))')
PYTHON_LIB=$(python3-config --configdir)

export CPLUS_INCLUDE_PATH=$PYTHON_INC

export CMAKE_PREFIX="/opt/homebrew/opt/lapack;/opt/homebrew/opt/openblas"

cmake .. -DPYTHON_EXECUTABLE=$PYTHON_EXEC -DPYTHON_LIBRARY=$PYTHON_LIB -DPYTHON_INCLUDE_DIR=$PYTHON_INC -DCMAKE_PREFIX_PATH=$CMAKE_PREFIX
make -j 8
make pygimli
```
After a `cd ..` I got the following result:
```
python -c 'import pygimli as pg; print(pg.__version__)'
1.4.6+43.gc2601998 (master)
```

I hope this addition might help.

Cheers,
Ralph

**Reply by florian-wagner** [reply]
Thanks for the feedback. We need to update the compilation instructions for Mac OS on the website.

**Comment by florian-wagner** [comment]
Dear @robinthibaut @ruboerner,

could you please try if `conda create -n pg15 -c gimli -c conda-forge pygimli=1.5.0` works on your Mac machines?

Thanks and best wishes
Florian

**Reply by robinthibaut** [reply]
Dear @florian-wagner,
The installation went smoothly, and I successfully ran an ERT inversion. Thanks!!
Robin

**Reply by florian-wagner** [reply]
Thanks Robin! Your feedback (and compilation instructions above) are much appreciated! If you have pyGIMLi compiled manually or if you simply install the `pgcore` package and use the git-sources for pyGIMLi as described here: https://www.pygimli.org/contrib.html#d-contribute-to-code You are very welcome to contribute a PR if you spot anything in the code or have developed something useful. We appreciate every little contribution (from typos over new examples to new functionality).

Stay in touch
Florian

**Reply by ruboerner** [reply]
Hi @florian-wagner,

I can also confirm that the installation works. It runs smoothly!

Thanks,
Ralph

**Comment by PavelShashkin** [comment]
Hi! Just installed with conda on MacBook Air M2 (python=3.9 and 3.11, pygimli=1.5.0 or 1.5.3). Installed smoothly. Trying to run 3D refraction example. Get "sh: tetgen: command not found" on mesh creation (both in spyder and jupyter notebook)- no idea where to go now...

**Reply by florian-wagner** [reply]
Sorry for the late reply, just saw this. Unfortunately, the `tetgen` packages in conda have not been updated since 2020 and are only provided for the older Mac models with intel chips: https://anaconda.org/channels/conda-forge/packages/tetgen/files?sort=-updated_at

So for May, `conda install pygimli` does not install tetgen. You have to provide it differently by compiling the binary. Or you try to install tetgen (and potentially also pygimli) via pip.

**Comment by AGeissler** [comment]
Hi all! 
Sadly, the instructions https://www.pygimli.org/compilation.html#sec-build did not work for me (M2 Mac, Sequoia 15.7.3). Everything seemed to work fine up to the 
`make -j 8`
command. This resulted in very 'deep-level' (IMO) errors from the compiler, e.g. 
`fatal error: 'cholmod.h' file not found` and
`error: invalid token at start of a preprocessor expression` for
 `#elif READPROC_FOUND || defined(HAVE_PROC_READPROC)`
pointing at the '||'. I was unable to find any hints on what may be the issue, the first one seems to be connected to the compile job not finding the / components of the SuiteSparse package. During the 'conda install', SuiteSparse and cholmod seem to be fine:

```-- Checking for package 'CHOLMOD'
-- SuiteSparse_config version: 7.10.1
-- SuiteSparse_config include: /Users/achim/Development/miniforge3/include/suitesparse`
-- SuiteSparse_config library: /Users/achim/Development/miniforge3/lib/libsuitesparseconfig.7.10.1.dylib
-- SuiteSparse_config static:  ```
-- AMD version: 3.3.3
-- AMD include: /Users/achim/Development/miniforge3/include/suitesparse
-- AMD library: /Users/achim/Development/miniforge3/lib/libamd.3.3.3.dylib
-- AMD static:  
-- Found CHOLMOD: /opt/homebrew/lib/libcholmod.dylib
-- CHOLMOD_LIBRARIES: /opt/homebrew/lib/libcholmod.dylib
```

Any ideas would be welcome.

One small issue I had with the instructions was `conda activate pygimli_env` followed by `conda install -c conda-forge boost numpy scipy matplotlib openblas suitesparse`. I had to deactivate the python environment again for conda install to work (didn't switch consoles).

Best,
Achim

**Reply by florian-wagner** [reply]
Hi Achim, sorry, just saw this. I think the problem is that suitesparse is taken from miniforge whereas cholmod is taken from the homebrew directory. 

Can you try: 

``` bash
export CMAKE_PREFIX=/Users/achim/Development/miniforge3

git clone https://github.com/gimli-org/pyGIMLi.git source

rm -rf build
mkdir build
cd build

cmake  \
    -DBLAS_openblas_LIBRARY=$CMAKE_PREFIX/lib/libopenblas.dylib \
    -D NOREADPROC=1 \
    -DUMFPACK_LIBRARIES=$CMAKE_PREFIX/lib/libumfpack.dylib \
    -DUMFPACK_INCLUDES=$CMAKE_PREFIX/include/suitesparse \
    -DCHOLMOD_LIBRARIES=$CMAKE_PREFIX/lib/libcholmod.dylib \
    -DCHOLMOD_INCLUDE_DIRS=$CMAKE_PREFIX/include/suitesparse \
    -DOpenBLAS_INCLUDE_DIR=$CMAKE_PREFIX/include \
    ../source

make -j 12 gimli
make -j 8 pygimli J=8 VERBOSE=1
cd ../source
unset PYTHONPATH
python -c 'import pygimli'
cd ..
```

Best wishes

Florian

**Reply by AGeissler** [reply]
Thanks! In the meantime, I have moved on to FEniCS and then FEniCSx/DOLFINx - I need to go back and see if PyGIMLI is still of interest to me (there are open issues with my current solutions/approaches). However, possibly PyGIMLi is somewhat overkill for the comparatively basic FE-Analysis I have in mind (transient thermal bridging in the building/construction sector). But I will probably give the installation another go, anyway, out of principle ;).

**Comment by PavelShashkin** [comment]
Hi, thank you for the reply!

Solution was to make tetgen manually - details is below if interecting

# 1. Download TetGen source code
Go to
https://codeberg.org/TetGen/TetGen

Download reasonable version
# 2. Extract the archive
tar -xzf tetgen1.5.0.tar.gz # Check the version you downloaded
cd tetgen1.5.0

# 3. Compile with Apple's clang (optimized for M2)
make

# 4. Install the binary to a location in your PATH
sudo cp tetgen /usr/local/bin/

# 5. Verify installation
tetgen -h

and then

PYTHON_EXEC=$(which python3)
PYTHON_INC=$(python3 -c 'import sysconfig; print(sysconfig.get_path("include"))')
PYTHON_LIB=$(python3-config --configdir)

export CPLUS_INCLUDE_PATH=$PYTHON_INC
export UMFPACK_ROOT="/opt/homebrew/Cellar/suite-sparse/7.12.1"
export EXTERNAL_DIR="/opt/homebrew/Cellar/suite-sparse/7.12.1"
export CMAKE_PREFIX_PATH="/opt/homebrew/opt/openblas"
cmake .. -DPYTHON_EXECUTABLE=$PYTHON_EXEC -DPYTHON_LIBRARY=$PYTHON_LIB -DPYTHON_INCLUDE_DIR=$PYTHON_INC
make -j 8
make pygimli


Cheers,
Pavel
________________________________

**Reply by florian-wagner** [reply]
Thanks Pavel, this can be helpful for future users.
